src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+            auto const original_function_call_instruction = *location;
+            *location = goto_programt::make_assertion(
+              false_exprt{},
+              make_function_pointer_restriction_assertion_source_location(
+                original_function_call_instruction.source_location,
+                restriction));
+            auto const assume_false_location = goto_function_body.insert_after(
+              location,
+              goto_programt::make_assumption(
+                false_exprt{},
+                original_function_call_instruction.source_location));
+            // this is mutable because we'll update this at the end of each
+            // loop iteration to always point at the start of the branch
+            // we created
+            auto else_location = location;
+            auto const end_if_location = goto_function_body.insert_after(
+              assume_false_location, goto_programt::make_skip());
+            for(auto const &restriction_target : restriction.second)
+            {
+              auto new_instruction = original_function_call_instruction;
+              // can't use get_function_call because that'll return a const ref
+              const symbol_exprt &function_pointer_target_symbol_expr =
+                goto_model.symbol_table.lookup_ref(restriction_target)
+                  .symbol_expr();
+              to_code_function_call(new_instruction.code).function() =
+                function_pointer_target_symbol_expr;
+              auto const goto_end_if_location =
+                goto_function_body.insert_before(
+                  else_location,
+                  goto_programt::make_goto(
+                    end_if_location,
+                    original_function_call_instruction.source_location));
+              auto const replaced_instruction_location =
+                goto_function.second.body.insert_before(
+                  goto_end_if_location, new_instruction);
+              else_location = goto_function.second.body.insert_before(
+                replaced_instruction_location,
+                goto_programt::make_goto(
+                  else_location,
+                  notequal_exprt{
+                    pointer_symbol,
+                    address_of_exprt{function_pointer_target_symbol_expr}}));
+            }
+          }
+        }
+      });
+  }
+}
+
+void parse_function_pointer_restriction_options_from_cmdline(
+  const cmdlinet &cmdline,
+  optionst &options)
+{
+  options.set_option(
+    RESTRICT_FUNCTION_POINTER_OPT,
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_OPT));
+  options.set_option(
+    RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT,
+    cmdline.get_values(RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT));
+}
+
+namespace
+{
+function_pointer_restrictionst::restrictionst
+merge_function_pointer_restrictions(
+  const function_pointer_restrictionst::restrictionst &lhs,
+  const function_pointer_restrictionst::restrictionst &rhs)
+{
+  auto result = lhs;
```

Comment:
:pick: Since you're immediately copying it, you could change `lhs` to be taken by value. That way if the user is doing

```
x = merge(x, y)
```
Then can do
```
x = merge(std::move(x), y)
```
and save a copy. 

It is also one less line of code!

Reply:
Done